### DISCUSSION TITLE
Using ROS2 from a Fprime component

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#1492] [Q&A] [answered]
**Objective**
Creating a **ROS2(foxy)** node inside my **Fprime 2.1** environment to be able to use ROS2 material for testing and visualization.

**First working architecture**

Inside my Fprime component called `InterfaceWithRos` I have a ROS2 node. In the `InterfaceWithRos.init()` I initialize ROS with `rclcpp::init(0, nullptr)` and I create a thread to run the `rclcpp::spin()`. Everything worked fine until this point. 

**Second NOT working architecture**

After the initial setup I decided to improve the architecture and instead of using `spin()` I changed to `spinOnce`  and instead of spinning on a different thread I decide to spin inside the port `SchedIn` of `InterfaceWithRos` connected to an `ActiveRateGroup`. In this way, the spin will be done only once for the tick of the `ActiveRateGroup`. I unit tested the full component creating test ROS nodes to check that the spin was working fine and the subscribers were receiving data from ROS.

The following step was to build my Fprime application and run it but here I started to encounter problems: the execution of  `spinOnce` create `Segmentation Fault`. 

**What I tried to understand the cause of the error:**
- Removing any node to be sure that the segmentation fault is not coming from my implementation of the node.
- Move the spin back to the `InterfaceWithRos.init()` function to check that the new changes impacted ROS in some way--> no errors.
- Move the `rclcpp::init(0, nullptr)` and spin inside the `SchedIn` (I know it does not make sense because the init should be done only once but it was done only to try to understand what was causing the error)--> The segmentation fault happen inside the `rclcpp::init(0, nullptr)`. 

**My observation:**
* Since everything works in the init and in the unit tests, I was thinking it is someway related to the `ActiveRateGroup` and/or the fact that the tick method is executed inside a thread created by `pthread`.
* A configuration of Fprime can impact the memory allocation of `rclcpp` causing the error.

**Questions**
- Do someone try to do something similar and had similar problems?
-  Any idea of what can cause the problem and/or suggestions to continue my journey in the ROS integration in Fprime?

**Comment by codeflight1** [comment]
I'm not sure what exactly is causing the error, but my first idea would be to move to an passive component with a custom thread, similar to how `TcpClient` works. That way you get complete control over the thread and you can simply put `spin` inside of it, along with any message handling code.

**Comment by SMorettini** [comment, marked as answer]
The Segmentation Fault was due to too-small stack size. In the Topology, I increased the stack size in the `ActiveRateGroup.start()`.
